mono/metadata/gc.c

```diff
@@ -216,7 +216,11 @@ mono_gc_run_finalize (void *obj, void *data)
 
 	runtime_invoke = domain->finalize_runtime_invoke;
 
-	mono_runtime_class_init (o->vtable);
+	exc = mono_runtime_class_init_full (o->vtable, FALSE);
+	if (exc) {
+		mono_unhandled_exception(exc);
+		exc = NULL;
```

Comment:
If the .cctor throws and exception, I think we want to handle and log like this, but then return. Maybe call `mono_domain_set_internal (caller_domain);` first like the code below and then `return;`